cmd/internal/backup/providers/s3/s3.go

```diff
@@ -189,49 +189,36 @@ func (b *BackupProviderS3) CleanupBackups(_ context.Context) error {
 }
 
 // DownloadBackup downloads the given backup version to the specified folder
-func (b *BackupProviderS3) DownloadBackup(ctx context.Context, version *providers.BackupVersion, outDir string) (string, error) {
+func (b *BackupProviderS3) DownloadBackup(ctx context.Context, version *providers.BackupVersion, writer io.Writer) error {
 	bucket := aws.String(b.config.BucketName)
 
-	downloadFileName := version.Name
-	if strings.Contains(downloadFileName, "/") {
-		downloadFileName = filepath.Base(downloadFileName)
-	}
-
-	backupFilePath := filepath.Join(outDir, downloadFileName)
-
-	f, err := b.fs.Create(backupFilePath)
-	if err != nil {
-		return "", err
-	}
-	defer f.Close()
-
 	downloader := s3manager.NewDownloader(b.sess)
 
-	_, err = downloader.DownloadWithContext(
+	buf := aws.NewWriteAtBuffer([]byte{})
```

Comment:
Unfortunately this loads the entire backup into memory, which we need to avoid. I am leaning towards a solution proposed [here](https://stackoverflow.com/questions/46019484/buffer-implementing-io-writerat-in-go), where the downloader is configured to use concurrency 1 and the existing writer is just wrapped into a `WriterAt` interface calling the regular write function. But maybe there are more ideas?

Reply:
I've done some research on this. I think its possible to implement something like a pipe with with WriteAt. See [here](https://pkg.go.dev/github.com/eikenb/pipeat)

Till now I will implement the sequential way.